Added support for max operating rate on all prefixes that support RFI

derflacco opened the pull request:
added c2.mtk to prefixes supporting RFI

Fix to achieve sub 10ms decoding latency for mtk g99 and MediaTekDimensity 8300 Ultra.
Tested with an Alldocube iplay 50 mini pro (hvec/h264), and Poco X6 pro (hvec).
Before i was getting about 15ms on the g99, and 20ms with the Poco.
Maybe can be useful for others mtk devices.

alonsojr1980 committed:
Added test flags for Mediatek SoCs

alonsojr1980 committed:
Added test flags for Mediatek SoCs

derflacco committed:
Added support for max operating rate on all prefixes that support RFI
added c2.mtk to prefixes supporting RFI

derflacco commented:
Testing @alonsojr1980 commits too, and seems to also reduce latency.

derflacco committed:
Experimental: Use more test flags found with the help of chatgpt.
-These flags may or may not work for every mtk device (or may not work at all, or partially work)
-I'm experimenting with them for fun, only to see if they actually do something, or reduce more the decoding latency
-In my personal testing, the flags do something, especially with my poco x6 pro (dimensity 8300 ultra), reducing the decoding to sub 6ms(, introducing a little audio distortion. My tablet with a g99 seems to gain little to nothing.
-The values are pretty thigh, they can introduce frame skipping, overheating, crash, tearing ecc, use them at your own risk

ClassicOldSong commented:
Thanks, I don't have recent MTK devices to test but there's an old one to play with.

I'll merge once I have confirmed the PR doesn't introduce regressions on other devices.

derflacco commented:
> Thanks, I don't have recent MTK devices to test but there's an old one to play with.
> 
> I'll merge once I have confirmed the PR doesn't introduce regressions on other devices.

Okay, thanks. I m posting my git around to search for additional testers. Seems that its working for other mtk chips too:
https://github.com/moonlight-stream/moonlight-android/issues/1241#issuecomment-3218651524
Let's see if its safe for other devices.

ClassicOldSong commented:
Just tried, it actually provides worse results on my G95 device... About 2-3ms slower than before

ClassicOldSong commented:
What's interesting is, the number goes lower when I use scrcpy, but the image gets more stutters

derflacco committed:
Reduce decoding latency on MTK:
 switch to non-blocking dequeue + “latest-frame” rendering, stabilize SurfaceView pacing
- attempt to further reduce decoding latency, results are promising (sub 5ms on both my mtk devices)
- add a decoder watchdog for C2 sleep recovery, to prevent decoder crashing
- switch to non-blocking dequeue + “latest-frame” rendering
- used chatgpt to help me with the code, feel free to clean and optimize it

set the trhead priority to Urget Display

- set the trhead priority to Urget Display
- reduce the dequeueOutputBuffer to 0

The goal here is to claim back some ms by "forcing" the decoding into big cores.
With my rooted Alldocube play 50 mini pro, i can gain more than 3ms by forcing the Top_App \ Foreground cpu_set to 6-7 (big cores only).
Performance governor should be used too, as it helps with less latency and stability
Of course i can't force those values without a rooted devices, so here i m just trying to gain some of the benefits.
Other settings that may help to achieve this goal are in MediaCodecHelper.java:

videoFormat.setInteger("vendor.mtk.vdec.thread.priority", 99);
videoFormat.setInteger("vendor.mtk.vdec.parser.boost", 1);
videoFormat.setInteger("vendor.mtk.vdec.cpu.boost.mode", 2);
videoFormat.setInteger("vendor.mtk.vdec.dvfs.mode", 1);

But in the last logging session, they seems to be skipped by my alldocube device entirely

derflacco commented:
> G95

G95 uses omx.mtk or c2.mtk decoder?

ClassicOldSong commented:
OMX.MTK

derflacco commented:
I will now try adding it to the frame invalidation support list. I hadn’t done it before because I don’t have a device with that decoder for testing. Brb with an apk build.

derflacco commented:
[app-nonRoot_game-debug.zip](https://github.com/user-attachments/files/21970328/app-nonRoot_game-debug.zip)
Here you go.

derflacco commented:
> What's interesting is, the number goes lower when I use scrcpy, but the image gets more stutters

Had the same problem with stock Moonlight/Artemis builds on my devices. When screen recording, I was able to reach sub-10ms decoding times (but with stutters). This behavior showed me that those chips were underutilized in some way and that they weren’t completely useless at decoding.

derflacco committed:
Add support for max operating rate for AVC too

derflacco committed:
added omx.mtk to prefixes supporting RFI
- untested, as i don't have a device using omx.mtk

derflacco committed:
Introduce: Latest-Frame Rendering
- Minimizes video delay by always showing the newest frame and dropping queued ones.
- May introduce micro stuttering on less powerful devices

derflacco committed:
Update README.md
Update README.md
Update README.md

derflacco committed:
Refactored MediaCodecDecoderRenderer:
- Measure decode latency (enqueue → dequeue)
- Remove the old uptime-based method
- Make dequeue timeout configurable via getOutputDequeueTimeoutUs()
- Replace hardcoded dequeueOutputBuffer

derflacco committed:
Unified decoder latency handling across all video pacing profiles

The timeout for getting decoded frames is now consistent everywhere, with a smarter backoff.

Frame presentation now uses proper timestamping, ensuring smoother vsync-aligned playback across Warp, Warp2, Latest-only, and Balanced profiles.

derflacco committed:
update strings
- Rename “Last Frame Rendering” → “LFR” across UI/docs
- Refresh latest-frame rendering section
- Update Ultra Low Latency description to include MTK

derflacco committed:
Add forceTightThresholds option to reduce latency

- New setting to apply stricter frame timing
- May improve responsiveness on some devices

ClassicOldSong commented:
Tried your latest experimental, it reduces the latency from 15ms to 12ms on the g95 with balanced frame pacing, 11ms with Warp2, made the device from unusable to usable

Great job!

I think we can move the two options to `Advanced settings` section and we're good to go, the ordinary Video settings part is way too crowded...

ClassicOldSong commented:
What's even more exciting is, with these options enabled, it's reporting 0.00ms decoding time on the 8Elite...

ClassicOldSong commented:
0.9ms on the 8sGen3

Impressive

derflacco committed:
use safeSet for Mtk flags
- add descriptions, format and clean code

derflacco committed:
add safeSet helper
- lets us to set codec params without crashing if a key isn’t supported
- code is cleaner: just call safeSet(fmt, key, value) instead of try/catch
- works for typical params like operating-rate, low-latency flags, custom vendor stuff

giftederic commented:
I tested my Samsung galaxy S10+ with Dimensity 9300+, with Artemis I was getting around 15~16 ms decoding time. Now with this experimental build from @derflacco published on Reddit I was getting around 7ms. So seem to help quite a bit for MTK chips.

ClassicOldSong commented:
https://github.com/ClassicOldSong/moonlight-android/releases/tag/v20.3.0-experimental.2

The experimental features is released in Artemis pre-release now

giftederic commented:
@ClassicOldSong i saw the newly added strings didn't have translation if I set the locale to Chinese. will it auto get translated? I can help with translating these string if needed. let me know how i can help or this is automatic. Appreciate your amazing work on Artemis! 🙏

ClassicOldSong commented:
It'll be translated upon stable release, or you can contribute to the translation, but the descriptions might change before stable release.

ClassicOldSong commented:
@derflacco when you're ready, tell me and I'll merge

I think you can safly move the other two improvements here already, it's pretty consistent (tested on different devices for hours withoud issue)

derflacco commented:
> @derflacco when you're ready, tell me and I'll merge
> 
> I think you can safly move the other two improvements here already, it's pretty consistent (tested on different devices for hours withoud issue)



> @derflacco when you're ready, tell me and I'll merge
> 
> I think you can safly move the other two improvements here already, it's pretty consistent (tested on different devices for hours withoud issue)

Hey! I completely forgot to check my phone.
I'm just polishing a few things and making some final adjustments. My Poco is running really stable, but I have a couple of small improvements to make on my G99.
I'll contact you as soon as everything is ready! 🙂

derflacco committed:
Merge branch 'moonlight-noir' of https://github.com/ClassicOldSong/moonlight-android into moonlight-noir

derflacco commented:
Ok, rebased + pushed. Cleaned up the commit layout, code untouched. Should be clearer now. Everything’s in the PR.

derflacco committed:
rename forceTightThresholds
- Strict Vsync

derflacco commented:
Hello, I’m not a big fan of the name forceTightThresholds in the UI. 

Here’s a commit I made with an alternative:

https://github.com/derflacco/moonlight-android/commit/664a7e143c716c5aff499982a39ac6ab95eb64d5

If you like it, feel free to merge this commit as well

derflacco committed:
build:
- add dependencies
- fix

derflacco committed:
UNTESED: better low-latency for Tegra + Snapdragon c2.qcom (snap 660,665 etc)

- added tegraDecoderPrefixes (omx.nvidia, c2.nvidia)
- skip LFR on NVIDIA (no need)
- set NVIDIA vendor keys: low-latency + no reordering
- expanded qualcommDecoderPrefixes with c2.qcom
- set Qualcomm vendor keys: low-latency, frame drop, disable reordering
- hook applyExtraVendorOptions() before configure()

derflacco committed:
UNTESED: better low-latency for Snapdragon Omx

- add Omx decoders (snap 865 etc)

- safeSet some vendor keys

derflacco committed:
disable frame reordering on OMX.qcom decoders

- In setDecoderLowLatencyOptions, stop forcing
  vendor.qti-ext-dec-picture-order.enable = 1
  (
- OMX.qcom set to 0 (disable reordering, lower latency)
- C2 (c2.qti/.qcom)  keep default behavior
- Also skip overwrite if the key was already set earlier

derflacco committed:
Mtk Prefixes fix
- Remove ULL/name gating
- Try both plain and .value variants via safeSet
- Expand mtkDecoderPrefixes (omx.mtk, c2.mtk, mtk, pentonic)
- Tune the Prefixes
- Try enabling different key variants along with their vendor/generic prefixes.
Unrecognized keys are ignored by the decoder, so it’s safe to set multiple options.
- Use AOSP hints (KEY_OPERATING_RATE=32767, KEY_PRIORITY=0)

derflacco committed:
Revert: Disable latest-only low latency path on NVIDIA decoders

derflacco committed:
ensure stable frame release under low-latency mode

Prevented CPU spin in low-latency mode by adding a small timeout instead of 0

Added a new release logic (“releaseWithPolicy”):

shows the frame immediately only if it’s almost in sync

otherwise keeps normal scheduling

safe fallback if anything goes wrong

Replaced old direct calls with the new safer logic

More robust handling when frameTimeNanos is missing (falls back to system clock)

Improved retry backoff → no more endless spin

derflacco committed:
Mtk prefixes:
- try also AOSP hints (KEY_OPERATING_RATE=32767, KEY_PRIORITY=0) even though we already apply them

derflacco committed:
fixup! Mtk Prefixes fix
- Some flags were causing decoder crashes, so use only the safest ones
- relax the values since we no longer use multiple profiles

derflacco committed:
Mtk prefixes: add vdec-lowlatency
- Enables low-latency decoding on older OMX.MTK CPUs
-Tested on Alldocube iPlay 30 tablet

derflacco committed:
Add Chart Overlay

Added LineChart views (latency / decode / FPS) with findViewById + visibility handling

Overlay container is now shown if charts are enabled, even with no text

Feeding data in onPerfUpdate with more robust parsing:

handles both comma and dot as decimal separator

Lite fallback: pattern "<latency>ms / <decode>ms" → takes the 2nd value

New addChartPoint(..) helper: creates dataset on demand, sliding window, invalidate()

Runnable calls qualified with Game.this.addChartPoint(..)

MediaCodecDecoderRenderer.java

Sends updates to overlay if (enablePerfOverlay || enablePerfCharts)

PreferenceConfiguration.java

Added new fields: enablePerfCharts, perfChartLatency, perfChartDecode, perfChartFps

Loads preferences from SharedPreferences

preferences.xml / strings.xml

New "Performance Charts" section with independent toggles

Added strings + contentDescription for latency/decode/fps charts

derflacco committed:
charts:
 - auto-enable Overlay Lite when enabling Charts; disable Charts if Overlay Lite is turned off
 - fix

ClassicOldSong commented:
@derflacco hey, can we slow the progress a bit, there're other PRs pending and I'm afraid your other modifications will conflict with them very soon at this pace... Can you migrate the latest branch with graphs to this commit and I can merge it to the next release?

derflacco commented:
@ClassicOldSong 
Do you mean the experimental branch?
Or should I push the latest one with charts as the last commit as moonlight-noir branch.
I'm not the most organized person lol

ClassicOldSong commented:
> Or should I push the latest one with charts as the last commit as moonlight-noir branch.

This

You can send a new PR for the FSR implementation when you think it's ready

derflacco commented:
to better understand:
you want:
moonlight noir up to (and including) commit bd6e925.
https://github.com/derflacco/moonlight-android/commits/experimental
or
Moonlight-Noir up to and including commit bd6e925,
 excluding commits 
 7efc7fc  through 85c455b
 https://github.com/derflacco/moonlight-android/commits/experimental

You can send a new PR for the FSR implementation when you think it's ready
This lives on its own branch

ClassicOldSong commented:
> moonlight noir up to (and including) commit https://github.com/ClassicOldSong/moonlight-android/commit/bd6e92576a7c03dcb048b7e3a5032ca3e1e9e87a.
https://github.com/derflacco/moonlight-android/commits/experimental

This

> This lives on its own branch

Since you force pushed a lot, it's pretty messy on the git timeline...

derflacco commented:
Yeah, I know, it’s a bad habit. Did the same thing back when I was messing around with custom Android kernels. Sorry about that!

ClassicOldSong commented:
Thank you for all the work!

derflacco commented:
thank you for the merge!

derflacco commented:
<img width="1920" height="1200" alt="Screenshot_20250906-000115_Diana" src="https://github.com/user-attachments/assets/219cf300-17e6-452d-9bd5-6eb27f5fa660" />
<img width="1920" height="1200" alt="Screenshot_20250906-000124_Diana" src="https://github.com/user-attachments/assets/fe946c8c-a7e2-4141-8677-3e4edaf339c0" />

Not sure this is the right place, but I tweaked charts curious what you think.

ClassicOldSong commented:
Actually I think they can be merged into one chart with different colors
